### DISCUSSION TITLE
Transparent background option <rh-audio-player>

### DISCUSSION THREAD

**Discussion by ocean-rh** [RedHat-UX/red-hat-design-system#1156] [Ideas]
### Description

<rh-audio-player> should have an option for a transparent background that could then pick up the background color of the div it lives in. And that would need to be combined with a light or dark theme for the controls in case the background color is dark. 

### Suggested solution

I would like another theme option built into the player component. 

### Example API

```HTML
<rh-audio-player color-palette="transparent-dark"></rh-audio-player>

or <rh-audio-player color-palette="transparent-light"></rh-audio-player>
```

**Comment by bennypowers** [comment]
Thanks for the report @ocean-rh. I'm going to convert this to a discussion until we prioritize this for design and implementation.

Some background on the color system:

The color-palette attribute takes a [set list of palette names](https://github.com/RedHat-UX/red-hat-design-system/blob/f79543b759c886561c2a6c771d655257495b8f0d/lib/context/color/provider.ts#L14-L31), so adding a new theme to player would mean adding a new theme to every other component.

As well, `<rh-audio-player>` has the [`--rh-audio-player-background-color` css custom property](https://github.com/RedHat-UX/red-hat-design-system/blob/f79543b759c886561c2a6c771d655257495b8f0d/elements/rh-audio-player/rh-audio-player.ts#L39-L44), which could be helpful in achieving a transparent player.

You can try this out by visiting https://ux.redhat.com/elements/audio-player/demo/customization/ and running this snippet in the browser's dev tools:

```js
$('main').style.backgroundColor = 'red';
$('rh-audio-player').style.setProperty('--rh-audio-player-background-color', 'transparent');
```

**Reply by ocean-rh** [reply]
The goal for this update is to avoid the need for additional css. We need the option for custom colors but having a transparent background seems like the easier option because we can then place it on the background we want. Most of those colors are already built in to whatever page they are on. We will be retrofitting several pages with the new player.

**Reply by bennypowers** [reply]
Can you paste here a few examples of the markup for pages you want to retrofit player on to, so we can get a better idea of the challenges involved?

**Reply by nikkimk** [reply]
What if we did a search for references to the old player's CSS (since you'd be retiring it once fully migrated anyway):

Find
 ```
.plyr-theme .plyr__controls {
    background-color: #633ec5
}
```

replace with 
```
.plyr-theme .plyr__controls {
    background-color: #633ec5;
}
rh-audio-player {
  --rh-audio-player-background-color: #633ec5;
}
```

so that you wouldn't need to add `transparent` to every instance of the player's HTML or add to the complexity of the component itself?

**Reply by ocean-rh** [reply]
My team is hoping to get the option to not need custom css in the future. That was the ask behind this request. I expect that it will still be needed in older instances and in special cases. 

All Command Line Heroes episode pages: https://www.redhat.com/en/command-line-heroes/season-9/relentless-replicators
Compiler: Hub and all interior pages: https://www.redhat.com/en/compiler-podcast
Code Comments: https://www.redhat.com/en/code-comments-podcast

**Reply by markcaron** [reply]
I'm hesitant to add a new transparent variant for a couple reasons:

1. Adding a variant (and extra JS overhead) to do something that CSS can do through CSS vars and `::part()` selectors we've exposed for this exact purpose.
2. Adding a transparent variant which shifts the liability on RHDS for accessibility compliance. Meaning, if we provide a transparent option, are we responsible for any contrast issues with the text and controls in Audio Player as a result of the transparency? I'd much prefer this "variance" to be fully owned by the implementer.

The preferred way to add transparency would be through a CSS `::part()` selector. Through this, one could make _only_ the toolbar transparent instead of potentially _all_ backgrounds in Audio Player (like the kebab menu).

This could be done at a global CSS level or in a style block on the page:
```
<style>
rh-audio-player::part(toolbar) {
    background: transparent;
}
</style>
```

Tighter control could be achieved in global CSS with a `[transparent]` attribute selector:
```
rh-audio-player[transparent]::part(toolbar) {
    background: transparent;
}
```

Or through a `.transparent` CSS class:
```
rh-audio-player.transparent::part(toolbar) {
    background: transparent;
}
```

Screenshot of the `part="toolbar"` we've exposed already for these situations:
<img width="1007" alt="image" src="https://github.com/RedHat-UX/red-hat-design-system/assets/1890950/55ee7234-dfef-48b3-85b2-739136b2936c">


The 2nd reason listed above, though, is what is most concerning, IMO. There's just no programatic way we can account for all the potential abuses of a transparent background. Depending on the background color—like highly saturated colors where neither white or black text works—the implementer may have to write their own CSS to ensure that the text and controls meet the required contrast guidelines anyway.

**Reply by markcaron** [reply]
@ocean-rh, I should note, that I'm not suggesting you'd "abuse" the transparent background, but that once we offer that option we open a can of worms for others to use in unexpected ways. I'd prefer that any customizations that break accessibility policies be 100% their fault.

![image](https://github.com/RedHat-UX/red-hat-design-system/assets/1890950/a9f8c4a2-469e-48a9-bb7a-2ff39dcb07cb)

**Comment by bennypowers** [comment]
Closing because rh-audio-player already supports OP's requirement and because we're not planning to add a one-off attribute, per our principle of preferring composition of primitive features to multiplying one-offs.
